File: 0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, right = 0;
        unordered_set<char> seen;
        int maxLength = 0;

        for (;right < s.size(); ++right) {
            if (seen.find(s[right]) != seen.end()) {
                while (left < right && seen.find(s[right]) != seen.end()) {
                    seen.erase(s[left]);
                    left++;
                }
            }

            seen.insert(s[right]);
            maxLength = max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/solution.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        lastSeen = dict()
        ans = 0
        j = 0
        for i, ch in enumerate(s):
            if ch in lastSeen:
                j = max(j, lastSeen[ch])
            ans = max(ans, i - j + 1)
            lastSeen[ch] = i+1
            
        return ans
